[FLINK-26727][python] Fix the implementation of sub-interpreter in Thread Mode (#19150)
apache/flink: docs/content.zh/docs/dev/python/python_execution_mode.md

Prior to release-1.15, there is the only execution mode called `PROCESS` execution mode. The `PROCESS`
mode means that the Python user-defined functions will be executed in separate Python processes.

In release-1.15, it has introduced a new execution mode called `THREAD` execution mode. The `THREAD`

> mode means that the Python user-defined functions will be executed in the same thread as Java Operator,

Reviewer: ```suggestion
mode means that the Python user-defined functions will be executed in the same process as Java Operator,
```

> but it will be affected by GIL performance.

Reviewer: ```suggestion
It should be noted that multiple Python user-defined functions running in the same JVM are still affected by GIL.
```

## When can/should I use THREAD execution mode?

The purpose of the introduction of `THREAD` mode is to overcome the overhead of serialization/deserialization
and network communication caused in `PROCESS` mode. So if performance is not your concern, or the computing
logic of your customized Python functions is the performance bottleneck of the job, `PROCESS` mode will
be the best choice as `PROCESS` mode provides the best isolation compared to `THREAD` mode.

## Configuring Python execution mode

The execution mode can be configured via the `python.execution-mode` setting.
There are two possible values:

 - `PROCESS`: The Python user-defined functions will be executed in separate Python process. (default)

>  - `THREAD`: The Python user-defined functions will be executed in the same thread as Java Operator.

Reviewer: ```suggestion
 - `THREAD`: The Python user-defined functions will be executed in the same process as the Java operator.
```

You could specify the Python execution mode using Python Table API as following:

```python
# Specify `PROCESS` mode
table_env.get_config().get_configuration().set_string("python.execution-mode", "process")

# Specify `THREAD` mode
table_env.get_config().get_configuration().set_string("python.execution-mode", "thread")
```

{{< hint info >}}
Currently, it still doesn't support to execute Python UDFs in `THREAD` execution mode in all places.
It will fall back to `PROCESS` execution mode in these cases. So it may happen that you configure a job

> to execute in `THREAD` execution modes, however, it's actually executed in `PROCESS` execution mode.

Reviewer: ```suggestion
to execute in `THREAD` execution mode, however, it's actually executed in `PROCESS` execution mode.
```

{{< /hint >}}
{{< hint info >}}

> `THREAD` execution mode only supports Python 3.7+.

Reviewer: ```suggestion
`THREAD` execution mode is only supported in Python 3.7+.
```
